seacatauth/client/service.py

```diff
@@ -0,0 +1,120 @@
+import logging
+import re
+import secrets
+
+import asab.storage.exceptions
+import asab.exceptions
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class ClientService(asab.Service):
+
+	ClientCollection = "cl"
+	ClientIdPattern = r"[a-z][a-z0-9._-]{2,31}"
+	ClientSecretLength = 32
+
+	def __init__(self, app, service_name="seacatauth.ClientService"):
+		super().__init__(app, service_name)
+		self.StorageService = app.get_service("asab.StorageService")
+		self.ClientIdRegex = re.compile("^{}$".format(self.ClientIdPattern))
+
+
+	async def initialize(self, app):
+		await super().initialize(app)
+		# TODO: Auto-create clients for Seacat UI etc.
+
+
+	async def list(self, page: int = 0, limit: int = None, query_filter: dict = None):
+		collection = self.StorageService.Database[self.ClientCollection]
+
+		if query_filter is None:
+			query_filter = {}
+		cursor = collection.find(query_filter)
+
+		cursor.sort("_c", -1)
+		if limit is not None:
+			cursor.skip(limit * page)
+			cursor.limit(limit)
+
+		resources = []
+		count = await collection.count_documents(query_filter)
+		async for resource_dict in cursor:
+			resources.append(resource_dict)
+
+		return {
+			"data": resources,
+			"count": count,
+		}
+
+
+	async def get(self, client_id: str):
+		data = await self.StorageService.get(self.ClientCollection, client_id)
+		return data
+
+
+	async def create(
+		self,
+		client_id: str,
+		base_url: str,
+		description: str = None,
+		scope: list = None,
+	):
+		# TODO: Generate client secret
```

Comment:
Yes, it should be generated and shown to the user exactly once, as a result of this call.

Reply:
Implemented